Sources/DistributedSystem/SyncCallManager.swift

```diff
@@ -6,64 +6,134 @@
             }
+        }
+    }
+
+    var logger: Logger {
+        loggerBox.value
+    }
+
+    init(_ loggerBox: Box<Logger>) {
+        self.loggerBox = loggerBox
+    }
 
+    private func resumeContinuation(_ continuation: CheckedContinuation<any DistributedSystem.SerializationRequirement, Error>, with result: inout ByteBuffer) {
+        guard let typeHintSize = result.readInteger(as: UInt16.self),
+              let typeHint = result.readString(length: Int(typeHintSize)) else {
+            logger.error("Invalid result received")
+            return
+        }
+
+        guard let type = _typeByName(typeHint) else {
+            continuation.resume(throwing: DistributedSystemErrors.error("Result with undefined type '\(typeHint)'"))
+            return
+        }
+
+        guard let type = type as? DistributedSystem.SerializationRequirement.Type else {
+            continuation.resume(throwing: DistributedSystemErrors.error("Result type '\(typeHint) does not conform to DistributedSystem.SerializationRequirement"))
+            return
+        }
+
+        result.withUnsafeReadableBytes { bytes in
             do {
-                try body(callID)
-            } catch {
-                lock.withLockVoid {
-                    continuations.removeValue(forKey: callID)
+                let value: Any = try type.init(fromSerializedBuffer: bytes)
+                if let value = value as? (any DistributedSystem.SerializationRequirement) {
+                    continuation.resume(returning: value)
+                } else {
+                    continuation.resume(throwing: DistributedSystemErrors.error("result with invalid type '\(typeHint)'"))
                 }
+            } catch {
                 continuation.resume(throwing: error)
             }
         }
+    }
+
+    func waitResult<T: DistributedSystem.SerializationRequirement>(_ callID: UInt64) async throws -> T {
+        logger.trace("waiting result for \(callID)")
+        let result = try await withCheckedThrowingContinuation { continuation in
+            let result: Optional<ByteBuffer>? = self.lock.withLock {
+                if let result = self.results.removeValue(forKey: callID) {
+                    return result
+                } else {
+                    self.continuations[callID] = continuation
+                    return nil
+                }
+            }
+
+            if let result {
+                if var result {
+                    resumeContinuation(continuation, with: &result)
+                } else {
+                    continuation.resume(throwing: DistributedSystemErrors.connectionLost)
+                }
+            }
+        }
+        logger.trace("got result for \(callID)")
 
         if let result = result as? T {
             return result
         } else {
-            throw DistributedSystemErrors.error("Invalid result type")
+            throw DistributedSystemErrors.error("Result does not conform to \(T.self)")
```

Comment:
Maybe we can make specific error from this one?
e.g. `DistributedSystemErrors.invalidResultType(type: Type)`

Reply:
Fixed.